File: internal/forge/forge.go
// Package forge holds what posting scan results to a code host looks like
// when it isn't GitHub-shaped: the shared vocabulary (inline comments,
// suggestion outcomes) and the diff-hunk arithmetic every host needs,
// because they all share the same constraint — an inline comment can only
// land on a line the diff contains.
//
// The host-specific packages (githubpr, gitlabmr) implement Forge. main
// picks one from the CI environment it finds itself in.
package forge

import (
	"strconv"
	"strings"
)

// InlineComment is one comment to attach to a line range of the diff, with
// Body already rendered in the host's own suggestion syntax — GitHub and
// GitLab both have one-click-applicable suggestion blocks, but the fence
// grammar differs, so rendering happens before this type is built.
type InlineComment struct {
	// Path is the file path relative to the repository root.
	Path string

	// StartLine and Line bound the commented range in the post-change file,
	// inclusive. A single-line comment sets them equal (or leaves StartLine
	// zero).
	StartLine int
	Line      int

	Body string

	// Marker uniquely identifies this comment's content. If a comment
	// already on the change contains it, this one is skipped — inline
	// comments can't be upserted as a set the way a summary comment can, so
	// recognizing one's own past comments is the only defence against
	// stacking duplicates on every push.
	Marker string
}

// SuggestionOutcome accounts for every comment handed to PostSuggestions.
// Nothing is dropped silently: a suggestion that never appears because its
// line isn't in the diff looks identical, from the outside, to a scanner
// that found nothing.
type SuggestionOutcome struct {
	Posted       int
	AlreadyThere int
	OutsideDiff  int
}

// Forge is a code host the scanner can report to.
type Forge interface {
	// UpsertComment finds the existing summary comment containing marker
	// and replaces its body, or creates one.
	UpsertComment(body, marker string) error

	// PostSuggestions attaches the comments as inline review comments,
	// filtering out anything the host would reject and anything already
	// posted. headSHA, when non-empty, pins the batch to the revision the
	// scan actually ran against.
	PostSuggestions(summary string, comments []InlineComment, headSHA string) (SuggestionOutcome, error)
}

// PatchLineNumbers walks a unified diff patch hunk by hunk and returns the
// set of post-change line numbers it covers — added and context lines both,
// since a finding on an unchanged resource header is anchored to a context
// line. Deleted lines have no position in the new file and are excluded.
func PatchLineNumbers(patch string) map[int]bool {
	lines := map[int]bool{}
	newLine := 0

	// The trailing newline would otherwise yield one phantom line past the
	// end of the last hunk, and a comment there is rejected by every host.
	for _, l := range strings.Split(strings.TrimSuffix(patch, "\n"), "\n") {
		if strings.HasPrefix(l, "@@") {
			if n, ok := hunkNewStart(l); ok {
				newLine = n
			}
			continue
		}
		if newLine == 0 {
			continue // text before the first hunk header
		}
		switch {
		case strings.HasPrefix(l, "+"), strings.HasPrefix(l, " "), l == "":
			// Added or unchanged: this line exists in the new file. An empty
			// string is an unchanged blank line whose leading space was
			// trimmed somewhere along the way.
			lines[newLine] = true
			newLine++
		case strings.HasPrefix(l, "-"):
			// Deleted: present only in the old file; the counter must not
			// advance.
		default:
			// "\ No newline at end of file" and anything else unrecognized.
		}
	}
	return lines
}

// hunkNewStart reads the post-change starting line out of a hunk header,
// e.g. "@@ -12,7 +14,9 @@ resource ..." -> 14.
func hunkNewStart(header string) (int, bool) {
	plus := strings.Index(header, "+")
	if plus < 0 {
		return 0, false
	}
	rest := header[plus+1:]
	end := strings.IndexAny(rest, ", ")
	if end < 0 {
		return 0, false
	}
	n, err := strconv.Atoi(rest[:end])
	if err != nil || n < 1 {
		return 0, false
	}
	return n, true
}

// LinesInDiff reports whether every line of the comment's range is
// commentable in the given per-file diff line sets.
func LinesInDiff(diffLines map[string]map[int]bool, cm InlineComment) bool {
	inFile, ok := diffLines[cm.Path]
	if !ok {
		return false
	}
	start := cm.StartLine
	if start <= 0 {
		start = cm.Line
	}
	for l := start; l <= cm.Line; l++ {
		if !inFile[l] {
			return false
		}
	}
	return true
}

File: src/tfpdf/forge.py
"""À quoi ressemble la publication de résultats de scan sur une forge quand elle n'a pas la forme
de GitHub."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Protocol


@dataclass(slots=True)
class InlineComment:
    """Un commentaire à rattacher à une plage de lignes du diff."""

    #: File path relative to the repository root.
    path: str

    #: Bound the commented range in the post-change file, inclusive. A
    #: single-line comment sets them equal (or leaves `start_line` zero).
    line: int
    start_line: int = 0

    body: str = ""

    #: Uniquely identifies this comment's content. If a comment already on the
    #: change contains it, this one is skipped — inline comments can't be
    #: upserted as a set the way a summary comment can, so recognizing one's
    #: own past comments is the only defence against stacking duplicates on
    #: every push.
    marker: str = ""


@dataclass(slots=True)
class SuggestionOutcome:
    """Rend compte de chaque commentaire confié à `post_suggestions`."""

    posted: int = 0
    already_there: int = 0
    outside_diff: int = 0


class Forge(Protocol):
    """Une forge à laquelle le scanner peut rapporter."""

    def upsert_comment(self, body: str, marker: str) -> None:
        """Trouve le commentaire de synthèse existant contenant `marker` et remplace
        son corps, ou en crée un."""
        ...

    def post_suggestions(
        self, summary: str, comments: list[InlineComment], head_sha: str
    ) -> SuggestionOutcome:
        """Rattache les commentaires en commentaires de revue en ligne, en écartant
        ce que l'hôte refuserait et ce qui a déjà été posté."""
        ...


def patch_line_numbers(patch: str) -> set[int]:
    """Parcourt un patch au format unifié, hunk par hunk, et rend les numéros de lignes d'après
    changement qu'il couvre."""
    lines: set[int] = set()
    new_line = 0

    # The trailing newline would otherwise yield one phantom line past the end
    # of the last hunk, and a comment there is rejected by every host.
    for line in patch.removesuffix("\n").split("\n"):
        if line.startswith("@@"):
            n = _hunk_new_start(line)
            if n is not None:
                new_line = n
            continue
        if new_line == 0:
            continue  # text before the first hunk header

        if line.startswith(("+", " ")) or line == "":
            # Added or unchanged: this line exists in the new file. An empty
            # string is an unchanged blank line whose leading space was
            # trimmed somewhere along the way.
            lines.add(new_line)
            new_line += 1
        elif line.startswith("-"):
            # Deleted: present only in the old file; the counter must not
            # advance.
            pass
        else:
            # "\ No newline at end of file" and anything else unrecognized.
            pass
    return lines


def _hunk_new_start(header: str) -> int | None:
    """La ligne de départ d'après changement d'un en-tête de hunk : par exemple
    « @@ -12,7 +14,9 @@ resource ... » donne 14."""
    plus = header.find("+")
    if plus < 0:
        return None
    rest = header[plus + 1 :]
    end = min((i for i in (rest.find(","), rest.find(" ")) if i >= 0), default=-1)
    if end < 0:
        return None
    try:
        n = int(rest[:end])
    except ValueError:
        return None
    return n if n >= 1 else None


def lines_in_diff(diff_lines: dict[str, set[int]], cm: InlineComment) -> bool:
    """Dit si chaque ligne de la plage du commentaire est commentable, au vu des
    ensembles de lignes de diff par fichier."""
    in_file = diff_lines.get(cm.path)
    if in_file is None:
        return False
    start = cm.start_line if cm.start_line > 0 else cm.line
    return all(line in in_file for line in range(start, cm.line + 1))
